### PR TITLE
Assignment 1 done

### PR DESCRIPTION
**Pull request by asish918**
# Assignment 1 Asish

### Drift Video Explanation
[Drift](https://video.drift.com/v/abYbUWWfUICx0mkXPwhrnN5nSnwEUHkb4XaUCyN1sF94/)

Test Case Screenshot -
Screenshot 2024-01-17 at 23 03 13

### CODE REVIEW

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 4 |
| code_smells | 49 |
| reliability_rating | D (at least 1 Critical Bug)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/assignments/1/src/main/java/org/example/ExecuteTransaction.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    public ExecuteTransaction() {
+
+    }
+
+    @Override
+    public void run() {
+        Transaction.Type type = null;
+
+        System.out.println(transaction.toString());
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/ExecuteTransaction.java:24]: Lets use loggers.

File: backend/assignments/1/src/main/java/org/example/parse/JSONParse.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Override
+    public String toString() {
+        return "Data{" +
+                "type='" + type + '\'' +
+                ", coinData=" + coinData +
+                '}';
+    }
+
+    public static class CoinData {
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/parse/JSONParse.java:45]: lets move this to another file.

File: backend/assignments/1/src/main/java/org/example/Main.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            String line;
+            br.readLine();
+            while ((line = br.readLine()) != null) {
+                String[] columns = line.split(",");
+
+                if(columns.length == 6) {
+                    Coin c = new Coin(Integer.parseInt(columns[1]), columns[2], columns[3], Double.parseDouble(columns[4]), Long.parseLong(columns[5]));
+                    Market.getCoinsList().add(c);
+                } else if (columns.length == 5) {
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/Main.java:51]: Separate this like TRADER_COLUMN_COUNT

File: backend/assignments/1/src/main/java/org/example/Main.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            Query.retrieveTopCoins(5);
+            Query.topBottomTraders();
+            Query.retrieveCoinDetails("LUNA");
+            Query.traderPortfolio("0x344427a90da861f79cc80bac2ff8638f");
+            Query.traderProfitLoss("0x344427a90da861f79cc80bac2ff8638f");
+        } catch (Exception e) {
+            Thread.currentThread().interrupt();
+        }
+    }
+    public static ArrayList<String[]> parseCSV(Path path) {
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/Main.java:39]: Pls add javadocs

File: backend/assignments/1/src/main/java/org/example/Main.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            parseCSV(Path.of(Constants.COIN_CSV_PATH));
+            parseCSV(Path.of(Constants.TRADER_CSV_PATH));
+            int transactionCount = jsonTransactions.size();
+            CountDownLatch latch = new CountDownLatch(transactionCount);
+            executeTransactions(jsonTransactions, latch);
+            latch.await();
+            Query.retrieveTopCoins(5);
+            Query.topBottomTraders();
+            Query.retrieveCoinDetails("LUNA");
+            Query.traderPortfolio("0x344427a90da861f79cc80bac2ff8638f");
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/Main.java:33]: Let's shift constants into a different file

File: backend/assignments/1/src/main/java/org/example/Main.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                }
+
+                data.add(columns);
+            }
+            return data;
+        } catch (IOException I) {
+            System.out.println(I);
+        }
+
+        return null;
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/Main.java:63]: Lets not return `null` here. Lets throw an exception and catch it.

File: backend/assignments/1/src/main/java/org/example/transaction/Transaction.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+import org.example.entities.Coin;
+import org.example.data.Market;
+import org.example.parse.JSONParse;
+import org.example.entities.Trader;
+
+import java.util.Objects;
+import java.util.Random;
+
+public class Transaction {
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/transaction/Transaction.java:11]: a better way could have been to make this class as an interface/parent class and implement/extend that

File: backend/assignments/1/src/main/java/org/example/transaction/Transaction.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        if(c == null) return;
+
+        c.increaseQuantity(coin.getVolume());
+        notifyAll();
+    }
+
+    public Trader findTrader(String walletAddress) {
+        for (Trader t : Market.getTraderList()) {
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/transaction/Transaction.java:102]: lets add validations for invalid `walletAddress`

File: backend/assignments/1/src/main/java/org/example/entities/Trader.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    public synchronized void removeCoins(Coin coin, long quantity) {
+        if (wallet.containsKey(coin)) {
+            long currentQuantity = wallet.get(coin);
+            if (currentQuantity >= quantity) {
+                wallet.put(coin, currentQuantity - quantity);
+                profitLoss += coin.getPrice() * quantity;
+            } else {
+                System.out.println("Error: Trying to sell more coins than owned by " + firstName + "  " + lastName);
```
Comment (rishav-kd) [inline backend/assignments/1/src/main/java/org/example/entities/Trader.java:84]: lets use loggers whenever possible.

**Review by rishav-kd — COMMENTED** [review]
Wrong Drift link. Other than that good overall implementation. Liked the modular approach. Make sure to use loggers and javadocs consistently and use validations whenever possible.

**Pull request closed without merging**